sycl/plugins/hip/pi_hip.hpp

```diff
@@ -370,18 +372,47 @@ struct _pi_mem {
 ///
 struct _pi_queue {
   using native_type = hipStream_t;
-
-  native_type stream_;
+  static constexpr int default_num_compute_streams = 64;
+  static constexpr int default_num_transfer_streams = 16;
+
+  std::vector<native_type> compute_streams_;
+  std::vector<native_type> transfer_streams_;
+  // delay_compute_ keeps track of which streams have been recently reused and
+  // their next use should be delayed. If a stream has been recently reused it
+  // will be skipped the next time it would be selected round-robin style. When
+  // skipped, its delay flag is cleared.
+  std::vector<bool> delay_compute_;
   _pi_context *context_;
   _pi_device *device_;
   pi_queue_properties properties_;
   std::atomic_uint32_t refCount_;
   std::atomic_uint32_t eventCount_;
-
-  _pi_queue(hipStream_t stream, _pi_context *context, _pi_device *device,
-            pi_queue_properties properties)
-      : stream_{stream}, context_{context}, device_{device},
-        properties_{properties}, refCount_{1}, eventCount_{0} {
+  std::atomic_uint32_t compute_stream_idx_;
+  std::atomic_uint32_t transfer_stream_idx_;
+  unsigned int num_compute_streams_;
+  unsigned int num_transfer_streams_;
+  unsigned int last_sync_compute_streams_;
+  unsigned int last_sync_transfer_streams_;
+  unsigned int flags_;
+  // When compute_stream_sync_mutex_ and compute_stream_mutex_ both need to be
+  // locked at the same time, compute_stream_sync_mutex_ should be locked first
+  // to avoid deadlocks
+  std::mutex compute_stream_sync_mutex_;
```

Comment:
It looks like `compute_stream_mutex_` is used  to guard access to the `compute_streams_`,  `transfer_stream_mutex_` is used  to guard access to the `transfer_streams_`. 
Could you please clarify what is purpose of `compute_stream_sync_mutex_` ? Probably it worth having a comment for it.
And should we have an additional mutex for `delay_compute_`?